students/guillaume/Q2_session02/integration.py

```diff
@@ -40,6 +40,77 @@ def sectors(*args):
     return values, bands
 
 
+def trapz_p(func, a, b, precision=10, **func_args):
+    '''
+    This function uses a more generic approach to trapeziodal numerical
+    integration in order to allow non constant size of the sub integration
+    sectors hence it is less efficient but more versatile
+    '''
+    # Test order of a & b
+    if a > b:
+        a, b = b, a
+
+    def loc_f(x):
+        nonlocal func_args
+        return func(x, **func_args)
+
+    x_l, f_x_l = sectors_precisions(loc_f, a, b, precision)
+
+    def band_f(lst, func):
+        l_v = len(lst)
+        return [func(x1, x2)
+                for x1, x2 in zip(islice(lst, 0, l_v - 1),
+                                  islice(lst, 1, l_v))]
+
+    band = band_f(x_l, lambda x1, x2: x2 - x1)
+    f_band = band_f(f_x_l, lambda x1, x2: 0.5 * (x1 + x2))
+    integral = sum([x * y for x, y in zip(band, f_band)])
+
+    return integral
+
+
+def sectors_precisions(*args):
+    '''
+    Assuming that f meets Riemann criterions for integration
+    default size of an integration sub domain is (1/20 x |b - a|)
+    for a sub dommain [x1, x2], where x2 >x 1
+    Z = |f(x2) - f(x1)| will be compared to a precision value
+    if Z bigger than precision x2 size will be decrease until the inequation
+    is accurate.
+    There is a test to avoid that calculations would run for too long
+    '''
+    func, a, b, precision = args
+
+    def prec(func, precision):
+
+        def alpha(x, step):
+            f_x = func(x), func(x + step)
+            results = x + step, f_x[1]
+            if (abs(f_x[1] - f_x[0]) <= precision) or step < precision / 100:
```

Comment:
are you sure this is the right criteria? If I'm reading right, this is measuring the magnitude of the change in f with the step. But if the function is linear, you can have a huge change (steep slope), and still get the exact answer.

So what you want is some measure of non-linearity -- how different is the function from an actual trapezoid?

you might check and see how different it is if you use two steps rather than one?

you also need to be careful about adding up too many really small numbers - you can get cancellation effects. Not a big deal with double precision floating point, but something to keep in mind.

Reply:
I build this criteria with the mindset of testing it on functions with various continuous parts but different definitions for each interval, I did not want to overkill so the calculation is relatively simple. Indeed linear function with a significant slope would be calculated very inefficiently. 
A very simple test on numerical derivation could help

Reply:
> A very simple test on numerical derivation could help

that's a good idea :-)